File: src/main/store/metadata.ts
import fs from 'fs'
import path from 'path'
import { DB_PATH, SIZZLE_CONFIG_DIR } from '../paths'

const TMP_PATH = path.join(SIZZLE_CONFIG_DIR, 'db.json.tmp')

interface ProjectMeta {
  lastLaunched: number | null
  tagOverride: ProjectTagOverride | null
  marker: ProjectMarker
}

export type ProjectMarker = 'favorite' | 'ignored' | null

export interface ProjectTag {
  name: string
  score: number
}

export interface ProjectTagOverride {
  tags: ProjectTag[]
  primaryTag: string | null
}

export interface ScanSettings {
  scanRoots: string[]
  ignoreRoots: string[]
  manualProjectRoots: string[]
}

interface DB {
  projects: Record<string, ProjectMeta>
  scanSettings?: ScanSettings
}

let cache: DB | null = null

function normalizeRootPath(rootPath: string): string {
  return path.resolve(rootPath.trim())
}

function sanitizeRootList(values: unknown): string[] {
  if (!Array.isArray(values)) return []
  const unique = new Set<string>()
  for (const value of values) {
    if (typeof value !== 'string') continue
    if (!value.trim()) continue
    const normalized = normalizeRootPath(value)
    if (normalized) unique.add(normalized)
  }
  return Array.from(unique)
}

function getNormalizedScanSettings(db: DB): ScanSettings {
  const scanRoots = sanitizeRootList(db.scanSettings?.scanRoots)
  const ignoreRoots = sanitizeRootList(db.scanSettings?.ignoreRoots)
  const manualProjectRoots = sanitizeRootList(db.scanSettings?.manualProjectRoots)
  return {
    scanRoots,
    ignoreRoots,
    manualProjectRoots,
  }
}

function ensureDir(): void {
  if (!fs.existsSync(SIZZLE_CONFIG_DIR)) {
    fs.mkdirSync(SIZZLE_CONFIG_DIR, { recursive: true })
  }
}

function readDB(): DB {
  if (cache) return cache
  ensureDir()
  try {
    const raw = fs.readFileSync(DB_PATH, 'utf-8')
    cache = JSON.parse(raw) as DB
  } catch {
    cache = { projects: {} }
  }
  return cache
}

function writeDB(db: DB): void {
  ensureDir()
  fs.writeFileSync(TMP_PATH, JSON.stringify(db, null, 2), 'utf-8')
  fs.renameSync(TMP_PATH, DB_PATH)
  cache = db
}

function pruneMissingProjects(db: DB): boolean {
  let changed = false
  for (const projectPath of Object.keys(db.projects)) {
    if (fs.existsSync(projectPath)) continue
    delete db.projects[projectPath]
    changed = true
  }
  return changed
}

export function getMetadata(projectPath: string): ProjectMeta {
  const db = readDB()
  const meta = db.projects[projectPath]
  if (!meta) return { lastLaunched: null, tagOverride: null, marker: null }
  return {
    ...meta,
    marker: normalizeMarker(meta.marker),
  }
}

export function setLastLaunched(projectPath: string): void {
  const db = readDB()
  if (!db.projects[projectPath]) {
    db.projects[projectPath] = { lastLaunched: null, tagOverride: null, marker: null }
  }
  db.projects[projectPath].lastLaunched = Date.now()
  writeDB(db)
}

export function getAllMetadata(): Record<string, ProjectMeta> {
  const db = readDB()
  const prunedMissingProjects = pruneMissingProjects(db)
  let normalized = prunedMissingProjects
  for (const projectPath of Object.keys(db.projects)) {
    if (!db.projects[projectPath].tagOverride) {
      db.projects[projectPath].tagOverride = null
      normalized = true
    }
    const normalizedMarker = normalizeMarker(db.projects[projectPath].marker)
    if (db.projects[projectPath].marker !== normalizedMarker) {
      db.projects[projectPath].marker = normalizedMarker
      normalized = true
    }
  }
  if (normalized) {
    writeDB(db)
  }
  return db.projects
}

function normalizeMarker(value: unknown): ProjectMarker {
  return value === 'favorite' || value === 'ignored' ? value : null
}

function sanitizeTagName(value: string): string {
  return value.trim().replace(/\s+/g, ' ')
}

function normalizeTagOverride(value: ProjectTagOverride): ProjectTagOverride {
  const tagMap = new Map<string, number>()
  for (const tag of value.tags) {
    if (!tag || typeof tag.name !== 'string') continue
    const name = sanitizeTagName(tag.name)
    if (!name) continue
    const score = Number.isFinite(tag.score) && tag.score > 0 ? tag.score : 0
    tagMap.set(name, Math.max(tagMap.get(name) ?? 0, score))
  }

  const tags = Array.from(tagMap.entries())
    .map(([name, score]) => ({ name, score }))
    .sort((a, b) => b.score - a.score || a.name.localeCompare(b.name))

  const totalScore = tags.reduce((sum, tag) => sum + tag.score, 0)
  const normalizedTags = totalScore > 0
    ? tags.map((tag) => ({ ...tag, score: tag.score / totalScore }))
    : tags.map((tag) => ({ ...tag, score: 1 / Math.max(tags.length, 1) }))

  const primaryTag = typeof value.primaryTag === 'string' && value.primaryTag.trim()
    ? sanitizeTagName(value.primaryTag)
    : null
  const resolvedPrimary = primaryTag && normalizedTags.some((tag) => tag.name === primaryTag)
    ? primaryTag
    : (normalizedTags[0]?.name ?? null)

  return {
    tags: normalizedTags,
    primaryTag: resolvedPrimary,
  }
}

export function setTagOverride(projectPath: string, override: ProjectTagOverride | null): ProjectMeta {
  const db = readDB()
  if (!db.projects[projectPath]) {
    db.projects[projectPath] = { lastLaunched: null, tagOverride: null, marker: null }
  }
  db.projects[projectPath].tagOverride = override ? normalizeTagOverride(override) : null
  db.projects[projectPath].marker = normalizeMarker(db.projects[projectPath].marker)
  writeDB(db)
  return db.projects[projectPath]
}

export function setProjectMarker(projectPath: string, marker: ProjectMarker): ProjectMeta {
  const db = readDB()
  if (!db.projects[projectPath]) {
    db.projects[projectPath] = { lastLaunched: null, tagOverride: null, marker: null }
  }
  db.projects[projectPath].marker = normalizeMarker(marker)
  writeDB(db)
  return db.projects[projectPath]
}

export function getScanSettings(): ScanSettings {
  const db = readDB()
  if (!db.scanSettings) {
    return {
      scanRoots: [],
      ignoreRoots: [],
      manualProjectRoots: [],
    }
  }
  const normalized = getNormalizedScanSettings(db)
  if (
    db.scanSettings.scanRoots?.length !== normalized.scanRoots.length
    || db.scanSettings.ignoreRoots?.length !== normalized.ignoreRoots.length
    || db.scanSettings.manualProjectRoots?.length !== normalized.manualProjectRoots.length
    || db.scanSettings.scanRoots?.some((value, index) => value !== normalized.scanRoots[index])
    || db.scanSettings.ignoreRoots?.some((value, index) => value !== normalized.ignoreRoots[index])
    || db.scanSettings.manualProjectRoots?.some((value, index) => value !== normalized.manualProjectRoots[index])
  ) {
    db.scanSettings = normalized
    writeDB(db)
  }
  return normalized
}

export function renameProjectMetadata(oldPath: string, newPath: string): void {
  const db = readDB()
  if (db.projects[oldPath]) {
    db.projects[newPath] = db.projects[oldPath]
    delete db.projects[oldPath]
    writeDB(db)
  }
}

export function setScanSettings(settings: ScanSettings): ScanSettings {
  const db = readDB()
  const normalized: ScanSettings = {
    scanRoots: sanitizeRootList(settings.scanRoots),
    ignoreRoots: sanitizeRootList(settings.ignoreRoots),
    manualProjectRoots: sanitizeRootList(settings.manualProjectRoots),
  }
  db.scanSettings = normalized
  writeDB(db)
  return normalized
}

File: src-tauri/src/commands/metadata.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use tauri::State;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectTag {
    pub name: String,
    pub score: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectTagOverride {
    pub tags: Vec<ProjectTag>,
    #[serde(rename = "primaryTag")]
    pub primary_tag: Option<String>,
}

pub type ProjectMarker = Option<String>;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectMeta {
    pub last_launched: Option<i64>,
    pub tag_override: Option<ProjectTagOverride>,
    pub marker: ProjectMarker,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScanSettings {
    pub scan_roots: Vec<String>,
    pub ignore_roots: Vec<String>,
    pub manual_project_roots: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct DB {
    projects: HashMap<String, ProjectMeta>,
    #[serde(default)]
    scan_settings: Option<ScanSettings>,
}

pub struct MetadataStore {
    config_dir: PathBuf,
    db_path: PathBuf,
    tmp_path: PathBuf,
    cache: Mutex<Option<DB>>,
}

impl MetadataStore {
    pub fn new(config_dir: PathBuf) -> Self {
        let db_path = config_dir.join("db.json");
        let tmp_path = config_dir.join("db.json.tmp");
        log::info!("[sizzle] MetadataStore db_path: {:?}", db_path);
        Self {
            config_dir,
            db_path,
            tmp_path,
            cache: Mutex::new(None),
        }
    }

    fn ensure_dir(&self) {
        fs::create_dir_all(&self.config_dir).ok();
    }

    fn read_db(&self) -> DB {
        let mut cache = self.cache.lock().unwrap();
        if let Some(ref db) = *cache {
            return db.clone();
        }
        self.ensure_dir();
        let db = match fs::read_to_string(&self.db_path) {
            Ok(raw) => serde_json::from_str(&raw).unwrap_or_else(|e| {
                log::info!("[sizzle] Failed to parse db.json: {}. First 500 chars: {}", e, &raw[..raw.len().min(500)]);
                DB {
                    projects: HashMap::new(),
                    scan_settings: None,
                }
            }),
            Err(e) => {
                log::info!("[sizzle] Failed to read db.json: {}", e);
                DB {
                    projects: HashMap::new(),
                    scan_settings: None,
                }
            },
        };
        log::info!("[sizzle] read_db: {} projects, scan_settings: {}",
            db.projects.len(),
            if db.scan_settings.is_some() { "present" } else { "none" }
        );
        *cache = Some(db.clone());
        db
    }

    fn write_db(&self, db: &DB) {
        self.ensure_dir();
        let json = serde_json::to_string_pretty(db).unwrap();
        fs::write(&self.tmp_path, &json).ok();
        fs::rename(&self.tmp_path, &self.db_path).ok();
        let mut cache = self.cache.lock().unwrap();
        *cache = Some(db.clone());
    }

    fn normalize_root_path(root: &str) -> String {
        Path::new(root.trim()).canonicalize().unwrap_or_else(|_| PathBuf::from(root.trim()))
            .to_string_lossy()
            .to_string()
    }

    fn sanitize_root_list(values: &[String]) -> Vec<String> {
        let mut seen = std::collections::HashSet::new();
        let mut result = Vec::new();
        for v in values {
            let trimmed = v.trim().to_string();
            if trimmed.is_empty() { continue; }
            let normalized = Self::normalize_root_path(&trimmed);
            if normalized.is_empty() { continue; }
            if seen.insert(normalized.clone()) {
                result.push(normalized);
            }
        }
        result
    }

    fn normalize_marker(marker: &ProjectMarker) -> ProjectMarker {
        match marker.as_deref() {
            Some("favorite") | Some("ignored") => marker.clone(),
            _ => None,
        }
    }

    fn normalize_tag_name(name: &str) -> String {
        name.trim().split_whitespace().collect::<Vec<_>>().join(" ")
    }

    fn normalize_tag_override(override_val: &ProjectTagOverride) -> ProjectTagOverride {
        let mut tag_map: HashMap<String, f64> = HashMap::new();
        for tag in &override_val.tags {
            let name = Self::normalize_tag_name(&tag.name);
            if name.is_empty() { continue; }
            let score = if tag.score.is_finite() && tag.score > 0.0 { tag.score } else { 0.0 };
            let entry = tag_map.entry(name).or_insert(0.0);
            *entry = entry.max(score);
        }

        let mut tags: Vec<ProjectTag> = tag_map
            .into_iter()
            .map(|(name, score)| ProjectTag { name, score })
            .collect();

        tags.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal)
            .then(a.name.cmp(&b.name)));

        let total_score: f64 = tags.iter().map(|t| t.score).sum();
        if total_score > 0.0 {
            for tag in &mut tags {
                tag.score /= total_score;
            }
        } else if !tags.is_empty() {
            let equal = 1.0 / tags.len() as f64;
            for tag in &mut tags {
                tag.score = equal;
            }
        }

        let primary_tag = override_val.primary_tag.as_ref()
            .map(|t| Self::normalize_tag_name(t))
            .filter(|t| !t.is_empty())
            .and_then(|t| {
                if tags.iter().any(|tag| tag.name == t) { Some(t) } else { None }
            })
            .or_else(|| tags.first().map(|t| t.name.clone()));

        ProjectTagOverride { tags, primary_tag }
    }

    pub fn get_metadata(&self, project_path: &str) -> ProjectMeta {
        let db = self.read_db();
        db.projects.get(project_path).cloned().unwrap_or(ProjectMeta {
            last_launched: None,
            tag_override: None,
            marker: None,
        })
    }

    pub fn set_last_launched(&self, project_path: &str) {
        let mut db = self.read_db();
        {
            let entry = db.projects.entry(project_path.to_string()).or_insert(ProjectMeta {
                last_launched: None,
                tag_override: None,
                marker: None,
            });
            entry.last_launched = Some(chrono::Utc::now().timestamp_millis());
        }
        self.write_db(&db);
    }

    pub fn get_all_metadata(&self) -> HashMap<String, ProjectMeta> {
        let db = self.read_db();
        log::info!("[sizzle] get_all_metadata: {} entries in db", db.projects.len());
        let mut changed = false;
        let mut result = db.projects.clone();

        let before = result.len();
        result.retain(|path, _| Path::new(path).exists());
        let retained = result.len();
        if retained != before {
            log::info!("[sizzle]   removed {} non-existent paths", before - retained);
        }

        for meta in result.values_mut() {
            if meta.tag_override.is_none() {
                // ensure it's None, not a weird value
            }
            let normalized = Self::normalize_marker(&meta.marker);
            if meta.marker != normalized {
                meta.marker = normalized;
                changed = true;
            }
        }

        if changed {
            let mut db = self.read_db();
            db.projects = result.clone();
            self.write_db(&db);
        }

        result
    }

    #[allow(dead_code)]
    fn trim_missing(&self, db: &mut DB) -> bool {
        let before = db.projects.len();
        db.projects.retain(|path, _| Path::new(path).exists());
        db.projects.len() != before
    }

    pub fn set_tag_override(&self, project_path: &str, override_val: Option<ProjectTagOverride>) -> ProjectMeta {
        let mut db = self.read_db();
        let result = {
            let entry = db.projects.entry(project_path.to_string()).or_insert(ProjectMeta {
                last_launched: None,
                tag_override: None,
                marker: None,
            });
            entry.tag_override = override_val.map(|o| Self::normalize_tag_override(&o));
            entry.marker = Self::normalize_marker(&entry.marker);
            entry.clone()
        };
        self.write_db(&db);
        result
    }

    pub fn set_project_marker(&self, project_path: &str, marker: ProjectMarker) -> ProjectMeta {
        let mut db = self.read_db();
        let result = {
            let entry = db.projects.entry(project_path.to_string()).or_insert(ProjectMeta {
                last_launched: None,
                tag_override: None,
                marker: None,
            });
            entry.marker = Self::normalize_marker(&marker);
            entry.clone()
        };
        self.write_db(&db);
        result
    }

    pub fn rename_project_metadata(&self, old_path: &str, new_path: &str) {
        let mut db = self.read_db();
        if let Some(meta) = db.projects.remove(old_path) {
            db.projects.insert(new_path.to_string(), meta);
            self.write_db(&db);
        }
    }

    pub fn get_scan_settings(&self) -> ScanSettings {
        let db = self.read_db();
        let settings = db.scan_settings.clone().unwrap_or(ScanSettings {
            scan_roots: vec![],
            ignore_roots: vec![],
            manual_project_roots: vec![],
        });

        log::info!("[sizzle] get_scan_settings: {} scan_roots, {} ignore_roots, {} manual_roots",
            settings.scan_roots.len(),
            settings.ignore_roots.len(),
            settings.manual_project_roots.len(),
        );
        if !settings.scan_roots.is_empty() {
            log::info!("[sizzle]   first scan_root: {:?}", settings.scan_roots[0]);
        }

        let normalized = ScanSettings {
            scan_roots: Self::sanitize_root_list(&settings.scan_roots),
            ignore_roots: Self::sanitize_root_list(&settings.ignore_roots),
            manual_project_roots: Self::sanitize_root_list(&settings.manual_project_roots),
        };

        // Write back if normalized differs
        if normalized != settings {
            let mut db = self.read_db();
            db.scan_settings = Some(normalized.clone());
            self.write_db(&db);
        }

        normalized
    }

    pub fn set_scan_settings(&self, settings: &ScanSettings) -> ScanSettings {
        let normalized = ScanSettings {
            scan_roots: Self::sanitize_root_list(&settings.scan_roots),
            ignore_roots: Self::sanitize_root_list(&settings.ignore_roots),
            manual_project_roots: Self::sanitize_root_list(&settings.manual_project_roots),
        };
        let mut db = self.read_db();
        db.scan_settings = Some(normalized.clone());
        self.write_db(&db);
        normalized
    }
}

// Tauri commands

#[tauri::command]
pub fn get_metadata(state: State<'_, MetadataStore>, project_path: String) -> ProjectMeta {
    state.get_metadata(&project_path)
}

#[tauri::command]
pub fn get_all_metadata(state: State<'_, MetadataStore>) -> HashMap<String, ProjectMeta> {
    state.get_all_metadata()
}

#[tauri::command]
pub fn set_last_launched(state: State<'_, MetadataStore>, project_path: String) {
    state.set_last_launched(&project_path);
}

#[tauri::command]
pub fn set_tag_override(
    state: State<'_, MetadataStore>,
    project_path: String,
    override_val: Option<ProjectTagOverride>,
) -> ProjectMeta {
    state.set_tag_override(&project_path, override_val)
}

#[tauri::command]
pub fn set_project_marker(
    state: State<'_, MetadataStore>,
    project_path: String,
    marker: ProjectMarker,
) -> ProjectMeta {
    state.set_project_marker(&project_path, marker)
}

#[tauri::command]
pub fn get_scan_settings(state: State<'_, MetadataStore>) -> ScanSettings {
    state.get_scan_settings()
}

#[tauri::command]
pub fn set_scan_settings(
    state: State<'_, MetadataStore>,
    settings: ScanSettings,
) -> ScanSettings {
    state.set_scan_settings(&settings)
}

#[tauri::command]
pub fn add_ignore_root(state: State<'_, MetadataStore>, root_path: String) -> ScanSettings {
    let mut current = state.get_scan_settings();
    current.ignore_roots.push(root_path);
    state.set_scan_settings(&current)
}
